src/segmenter/utils.rs

```diff
@@ -0,0 +1,68 @@
+use fst::raw::{Fst, Output};
+
+/// Final-state-transducer (FST) Segmenter
+pub(crate) struct FstSegmenter<'o> {
+    words_fst: &'o Fst<&'o [u8]>
+}
+
+impl<'o> FstSegmenter<'o> {
+    pub(crate) fn new(words_fst: &'o Fst<&'o [u8]>) -> Self {
+        Self { words_fst }
+    }
+}
+
+impl<'o> FstSegmenter<'o> {
+    pub fn segment_str(&'o self, mut to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o> {
```

Comment:
I would've like `FstSegmenter` to implement the `Segmenter` trait, e.g.
```rs
impl<'o> Segmenter for FstSegmenter<'o> {
```
but because the `FstSegmenter` struct has its own lifetime (to keep the FST alive for the lifetime of the struct), and the `Segmenter::segment_str` function signature also has its own lifetime:
https://github.com/meilisearch/charabia/blob/2beae7e18293f3de023181cd9dac4aab759b7a34/src/segmenter/mod.rs#L145-L148
I'm not sure how to do so without modifying the `Segmenter` trait (I'm relatively new to Rust).

If I use an anonymous lifetime, e.g.
```rs
impl Segmenter for FstSegmenter<'_> {
    fn segment_str<'o>(&self, mut to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o> {
```
it fails with:
```
error: lifetime may not live long enough
  --> src/segmenter/utils.rs:38:9
   |
17 |     fn segment_str<'o>(&self, mut to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o> {
   |                    --  ----- has type `&FstSegmenter<'1>`
   |                    |
   |                    lifetime `'o` defined here
...
38 |         Box::new(iter)
   |         ^^^^^^^^^^^^^^ associated function was supposed to return data with lifetime `'o` but it is returning data with lifetime `'1`
```

Reply:
Good question,
first, in this codebase, the lifetime `'o` is related to the **original** text, for example, the [`Token` has the lifetime `'o`](https://github.com/meilisearch/charabia/blob/6d0a01d371088283d4e9bbadba24c0a0d2344571/src/token.rs#L35-L53) because it keeps a reference of the **original** text via the field `lemma`.

In your case, the struct `FstSegmenter` doesn't store any references to the **original** text, so you shouldn't name your lifetime `'o` but `'d` or `'wd` for **word dictionary** or even `'fst`.

But, does this will fix your issue? I'm not completely sure, because [from_fn](https://doc.rust-lang.org/std/iter/fn.from_fn.html) struggles to figure out which lifetime is returned, and you would probably have to precise it:
```rust
let iter = std::iter::from_fn::<&'o str, _>(move || {
```
But I'm not 100% sure that would fix your issue 🤔

Reply:
Ah, I wasn't sure what the `'o` meant, thanks for clarifying it, I've renamed it to `'fst` as you suggested.

Reply:
The lifetime name is not the cause of the issue though, even after renaming it, it still fails with the same `lifetime may not live long enough` error.

```rs
impl<'fst> Segmenter for FstSegmenter<'fst> {
    fn segment_str<'o>(&self, mut to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o> {
        let iter = std::iter::from_fn::<&'o str, _>(move || {
```
fails with the same error as well. `<&'o str, _>` is exactly what is inferred. The issue is not that `from_fn` cannot figure out which lifetime is returned, but that `'o` might outlive `'fst`, since `'o` is determined by the caller.

We also cannot declare `'o` as a subtype of `fst` (to ensure that `'fst` outlives `'o`), e.g.
```rs
impl<'fst> Segmenter for FstSegmenter<'fst> {
    fn segment_str<'o: 'fst>(&self, mut to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o> {
```
because if the caller uses the `Segmenter` trait, then there is no guarantee that the `'o` lifetime will outlive `'fst`, since the `Segmenter` trait doesn't have the `'fst` lifetime.

So it seems that the only way for `FstSegmenter` to implement `Segmenter` is to modify the `Segmenter` trait, by moving the lifetime from the function to the trait, i.e.
```rs
pub trait Segmenter<'o>: Sync + Send {
    fn segment_str(&self, s: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o>;
}
```
Unfortunately, that would require changing all implementations of `Segmenter`, therefore I think its better for `FstSegmenter` to not implement `Segmenter`.